_posts/2020-01-21-graphql-relay-windowed-pagination.markdown

```diff
@@ -0,0 +1,328 @@
+helpers to make sure types and resolvers are compatible with some Relay expectations. So, we use this library to
+generate our cursors, and can convert the cursor to an offset. A page of 4 with a size of 10, returns the elements
+numbered 30 - 39 in that list. So a page of 4 (and size of 10), is equivalent to an offset of 29 (and size of 10).
+We have:
+
+```js
+const pageToCursor = (page, size) => {
+  return String((page - 1) * size - 1)
+}
+```
+
+This gives us the offset of the last value of the previous page. While our upstream services are all still
+paginating using this size/offset method, the [GraphQL cursor spec](https://relay.dev/graphql/connections.htm)
+prefers opaque cursors to be used on the client. This allows the actual implementation of pagination to change
+upstream while clients remain unaffected. Thus if we ever update our upstream pagination arguments/logic/setup, we
+could update this schema implementation accordingly, and clients would continue to be functional.
+
+For inspiration in constructing our `first`, `last`, and `around` groups, we turn to
+[Fingertips](https://www.fngtps.com/) and their
+[pagination library](https://github.com/Fingertips/peiji-san/blob/6bd1bc7c152961dcde376a8bcb2ca393b5b45829/lib/peiji_san/view_helper.rb#L87).
+That code goes through the various cases possible (a short list, a long list where the current page is near the
+front, middle or end, various degenerate cases, etc.), and returns a proper structure that represents this data. It
+can handle all combinations of list sizes, and current position relative to the total size.
+
+In pseudo-code, it looks like:
+
+```js
+if emptyList
+  around = [1]
+else if listIsShort
+  around = [1...totalPages]
+else if nearBeginning
+  around = [1...3]
+  last = [totalPages]
+else if nearMiddle
+  first = [1]
+  middle = [currentPage-1, currentPage, currentPage+1]
+  last = [totalPages]
+else if nearEnd
+  first = [1]
+  around = [last-1, last, last+1]
+```
+
+Our full implementation of that method can be found
+[here](https://github.com/artsy/metaphysics/blob/205592be7f59970cf80313972ceb95bb1579c31f/src/schema/v2/fields/pagination.ts#L96).
+
+For a real-life example, check out
+[this link, corresponding to a page number of 4](<https://metaphysics-staging.artsy.net/v2?query=%7B%0AartworksConnection(first%3A5%2C%20after%3A%20%22YXJyYXljb25uZWN0aW9uOjE0%22)%20%7B%0A%20%20pageInfo%20%7B%0A%20%20%20%20hasNextPage%0A%20%20%20%20endCursor%0A%20%20%7D%0A%20%20pageCursors%7B%0A%20%20%20%20first%20%7B%0A%20%20%20%20%20%20cursor%0A%20%20%20%20%20%20page%0A%20%20%20%20%20%20isCurrent%0A%20%20%20%20%7D%0A%20%20%20%20last%20%7B%0A%20%20%20%20%20%20cursor%0A%20%20%20%20%20%20page%0A%20%20%20%20%20%20isCurrent%0A%20%20%20%20%7D%0A%20%20%20%20around%20%7B%0A%20%20%20%20%20%20cursor%0A%20%20%20%20%20%20page%0A%20%20%20%20%20%20isCurrent%0A%20%20%20%20%7D%0A%20%20%20%20previous%20%7B%0A%20%20%20%20%20%20page%0A%20%20%20%20%20%20cursor%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%7D>).
```

Comment:
I'm not crazy about sending _any_ public traffic to this, but especially to a staging service. I understand the idea of avoiding production but... seems we're in a bad spot either way. 🤔